File: packages/arena-cli/bin/arena-cli.js
#!/usr/bin/env node

const DEFAULT_BASE = "https://agent-arena-rps.vercel.app";
const MOVES = ["ROCK", "PAPER", "SCISSORS"];

function parseArgs(argv) {
  const opts = { _: [] };
  for (let i = 0; i < argv.length; i += 1) {
    const a = argv[i];
    if (a.startsWith("--")) {
      const key = a.slice(2);
      const val = argv[i + 1] && !argv[i + 1].startsWith("--") ? argv[++i] : true;
      opts[key] = val;
    } else {
      opts._.push(a);
    }
  }
  return opts;
}

function usage() {
  console.log(`Agent Arena CLI

Usage:
  arena-cli <command> [options]

Commands:
  register      --name <name> [--email <email>] [--description <text>]
  qual-start    --key <apiKey> [--difficulty easy|medium|hard]
  qual-round    --key <apiKey> --qual-id <id> --round <n> --move ROCK|PAPER|SCISSORS
  qual-auto     --key <apiKey> [--difficulty easy|medium|hard] [--strategy paper|rock|scissors|random|cycle] [--max-rounds 5]
  join          --key <apiKey>
  queue
  queue-me      --key <apiKey>
  onboard       --name <name> [--email <email>] [--description <text>] [--difficulty easy|medium|hard] [--strategy ...]
  watch-lobby   [--interval 5]

Global:
  --base <url>  API base (default: ${DEFAULT_BASE})
`);
}

async function req(base, method, path, { key, body } = {}) {
  const headers = { accept: "application/json" };
  if (body) headers["content-type"] = "application/json";
  if (key) headers["x-agent-key"] = key;

  const res = await fetch(`${base.replace(/\/$/, "")}${path}`, {
    method,
    headers,
    body: body ? JSON.stringify(body) : undefined,
  });

  const text = await res.text();
  let json;
  try {
    json = text ? JSON.parse(text) : {};
  } catch {
    json = { raw: text };
  }

  if (!res.ok) {
    const err = new Error(`HTTP ${res.status}`);
    err.status = res.status;
    err.payload = json;
    throw err;
  }

  return json;
}

function chooseMove(strategy, roundNo) {
  const s = (strategy || "paper").toLowerCase();
  if (["rock", "paper", "scissors"].includes(s)) return s.toUpperCase();
  if (s === "random") return MOVES[Math.floor(Math.random() * MOVES.length)];
  return MOVES[(roundNo - 1) % 3];
}

function out(x) {
  console.log(JSON.stringify(x, null, 2));
}

async function main() {
  const argv = process.argv.slice(2);
  if (!argv.length || argv.includes("-h") || argv.includes("--help")) {
    usage();
    process.exit(0);
  }

  const cmd = argv[0];
  const opts = parseArgs(argv.slice(1));
  const base = opts.base || DEFAULT_BASE;

  try {
    if (cmd === "register") {
      if (!opts.name) throw new Error("--name is required");
      const body = { name: opts.name };
      if (opts.email) body.authorEmail = opts.email;
      if (opts.description) body.description = opts.description;
      out(await req(base, "POST", "/api/agents", { body }));
      return;
    }

    if (cmd === "qual-start") {
      if (!opts.key) throw new Error("--key is required");
      const body = { difficulty: opts.difficulty || "easy" };
      out(await req(base, "POST", "/api/agents/me/qualify", { key: opts.key, body }));
      return;
    }

    if (cmd === "qual-round") {
      if (!opts.key || !opts["qual-id"] || !opts.round || !opts.move) {
        throw new Error("--key --qual-id --round --move are required");
      }
      const path = `/api/agents/me/qualify/${opts["qual-id"]}/rounds/${opts.round}`;
      out(await req(base, "POST", path, { key: opts.key, body: { move: String(opts.move).toUpperCase() } }));
      return;
    }

    if (cmd === "qual-auto") {
      if (!opts.key) throw new Error("--key is required");
      const start = await req(base, "POST", "/api/agents/me/qualify", {
        key: opts.key,
        body: { difficulty: opts.difficulty || "easy" },
      });
      const qualId = start.qualMatchId;
      const maxRounds = Number(opts["max-rounds"] || 5);
      const strategy = opts.strategy || "paper";
      const rounds = [];
      let result = "IN_PROGRESS";
      for (let r = 1; r <= maxRounds; r += 1) {
        const move = chooseMove(strategy, r);
        const rr = await req(base, "POST", `/api/agents/me/qualify/${qualId}/rounds/${r}`, {
          key: opts.key,
          body: { move },
        });
        rounds.push(rr);
        result = rr.qualStatus || result;
        if (result === "PASSED" || result === "FAILED") break;
      }
      out({ qualMatchId: qualId, result, rounds });
      process.exit(result === "PASSED" ? 0 : 2);
    }

    if (cmd === "join") {
      if (!opts.key) throw new Error("--key is required");
      out(await req(base, "POST", "/api/queue", { key: opts.key }));
      return;
    }

    if (cmd === "queue") {
      out(await req(base, "GET", "/api/queue"));
      return;
    }

    if (cmd === "queue-me") {
      if (!opts.key) throw new Error("--key is required");
      out(await req(base, "GET", "/api/queue/me", { key: opts.key }));
      return;
    }

    if (cmd === "onboard") {
      if (!opts.name) throw new Error("--name is required");
      const regBody = { name: opts.name };
      if (opts.email) regBody.authorEmail = opts.email;
      if (opts.description) regBody.description = opts.description;

      const reg = await req(base, "POST", "/api/agents", { body: regBody });
      const key = reg.apiKey;
      const start = await req(base, "POST", "/api/agents/me/qualify", {
        key,
        body: { difficulty: opts.difficulty || "easy" },
      });
      const qualId = start.qualMatchId;
      const maxRounds = Number(opts["max-rounds"] || 5);
      const strategy = opts.strategy || "paper";
      const rounds = [];
      let result = "IN_PROGRESS";
      for (let r = 1; r <= maxRounds; r += 1) {
        const move = chooseMove(strategy, r);
        const rr = await req(base, "POST", `/api/agents/me/qualify/${qualId}/rounds/${r}`, {
          key,
          body: { move },
        });
        rounds.push(rr);
        result = rr.qualStatus || result;
        if (result === "PASSED" || result === "FAILED") break;
      }

      let queue = null;
      if (result === "PASSED") {
        queue = await req(base, "POST", "/api/queue", { key });
      }

      out({ register: reg, qualify: { qualMatchId: qualId, result, rounds }, queue });
      process.exit(result === "PASSED" ? 0 : 3);
    }

    if (cmd === "watch-lobby") {
      const interval = Number(opts.interval || 5);
      console.log(`Watching ${base} every ${interval}s... Ctrl+C to stop`);
      while (true) {
        const q = await req(base, "GET", "/api/queue");
        const m = await req(base, "GET", "/api/matches");
        out({
          ts: Date.now(),
          queue: q.queue || [],
          total: q.total ?? (q.queue || []).length,
          running: (m.matches || []).filter((x) => x.status === "RUNNING"),
          recent: (m.matches || []).filter((x) => x.status === "FINISHED").slice(0, 5),
        });
        await new Promise((r) => setTimeout(r, interval * 1000));
      }
    }

    throw new Error(`Unknown command: ${cmd}`);
  } catch (err) {
    out({
      error: err.message || "CLI_ERROR",
      status: err.status,
      details: err.payload,
    });
    process.exit(10);
  }
}

main();

File: tools/arena-cli.py
#!/usr/bin/env python3
"""
Agent Arena CLI (API-first)

No browser needed. Uses HTTPS API directly.

Examples:
  python3 tools/arena-cli.py register --name MyBot-001
  python3 tools/arena-cli.py qualify-auto --key ak_live_xxx
  python3 tools/arena-cli.py join --key ak_live_xxx
  python3 tools/arena-cli.py onboard --name MyBot-001
  python3 tools/arena-cli.py queue
  python3 tools/arena-cli.py queue-me --key ak_live_xxx
"""

from __future__ import annotations

import argparse
import json
import random
import time
import urllib.error
import urllib.request
from typing import Any, Dict, Optional

DEFAULT_BASE = "https://agent-arena-rps.vercel.app"
MOVES = ["ROCK", "PAPER", "SCISSORS"]


class ApiException(Exception):
    def __init__(self, status: int, payload: Any):
        self.status = status
        self.payload = payload
        super().__init__(f"API {status}: {payload}")


def _request(
    base: str,
    method: str,
    path: str,
    api_key: Optional[str] = None,
    body: Optional[Dict[str, Any]] = None,
    timeout: int = 20,
) -> Any:
    url = base.rstrip("/") + path
    data = None
    headers = {"accept": "application/json"}

    if body is not None:
        data = json.dumps(body).encode("utf-8")
        headers["content-type"] = "application/json"

    if api_key:
        headers["x-agent-key"] = api_key

    req = urllib.request.Request(url=url, data=data, headers=headers, method=method.upper())

    try:
        with urllib.request.urlopen(req, timeout=timeout) as resp:
            raw = resp.read().decode("utf-8")
            if not raw:
                return {"ok": True}
            try:
                return json.loads(raw)
            except json.JSONDecodeError:
                return {"raw": raw}
    except urllib.error.HTTPError as e:
        raw = e.read().decode("utf-8") if e.fp else ""
        payload: Any
        try:
            payload = json.loads(raw) if raw else {"error": f"HTTP_{e.code}"}
        except json.JSONDecodeError:
            payload = {"error": f"HTTP_{e.code}", "raw": raw}
        raise ApiException(e.code, payload) from e


def _print(obj: Any) -> None:
    print(json.dumps(obj, ensure_ascii=False, indent=2))


def _choose_move(strategy: str, round_no: int) -> str:
    s = strategy.lower()
    if s in ("rock", "paper", "scissors"):
        return s.upper()
    if s == "random":
        return random.choice(MOVES)
    return MOVES[(round_no - 1) % 3]


def cmd_register(args: argparse.Namespace) -> int:
    body: Dict[str, Any] = {"name": args.name}
    if args.email:
        body["authorEmail"] = args.email
    if args.description:
        body["description"] = args.description

    res = _request(args.base, "POST", "/api/agents", body=body)
    _print(res)
    return 0


def cmd_qual_start(args: argparse.Namespace) -> int:
    body: Dict[str, Any] = {}
    if args.difficulty:
        body["difficulty"] = args.difficulty
    res = _request(args.base, "POST", "/api/agents/me/qualify", api_key=args.key, body=body)
    _print(res)
    return 0


def cmd_qual_round(args: argparse.Namespace) -> int:
    path = f"/api/agents/me/qualify/{args.qual_id}/rounds/{args.round_no}"
    res = _request(args.base, "POST", path, api_key=args.key, body={"move": args.move.upper()})
    _print(res)
    return 0


def cmd_qual_auto(args: argparse.Namespace) -> int:
    start_body: Dict[str, Any] = {}
    if args.difficulty:
        start_body["difficulty"] = args.difficulty

    start = _request(args.base, "POST", "/api/agents/me/qualify", api_key=args.key, body=start_body)
    qual_id = start.get("qualMatchId")
    if not qual_id:
        _print({"step": "qualify-start", "response": start})
        return 1

    history = []
    for r in range(1, args.max_rounds + 1):
        move = _choose_move(args.strategy, r)
        path = f"/api/agents/me/qualify/{qual_id}/rounds/{r}"
        res = _request(args.base, "POST", path, api_key=args.key, body={"move": move})
        history.append(res)
        status = str(res.get("qualStatus", "")).upper()
        if status in ("PASSED", "FAILED"):
            _print({"qualMatchId": qual_id, "result": status, "rounds": history})
            return 0 if status == "PASSED" else 2

    _print({"qualMatchId": qual_id, "result": "IN_PROGRESS", "rounds": history})
    return 3


def cmd_join(args: argparse.Namespace) -> int:
    res = _request(args.base, "POST", "/api/queue", api_key=args.key)
    _print(res)
    return 0


def cmd_queue(args: argparse.Namespace) -> int:
    res = _request(args.base, "GET", "/api/queue")
    _print(res)
    return 0


def cmd_queue_me(args: argparse.Namespace) -> int:
    res = _request(args.base, "GET", "/api/queue/me", api_key=args.key)
    _print(res)
    return 0


def cmd_onboard(args: argparse.Namespace) -> int:
    reg_body: Dict[str, Any] = {"name": args.name}
    if args.email:
        reg_body["authorEmail"] = args.email
    if args.description:
        reg_body["description"] = args.description

    reg = _request(args.base, "POST", "/api/agents", body=reg_body)
    key = reg.get("apiKey")
    if not key:
        _print({"step": "register", "response": reg})
        return 1

    start_body: Dict[str, Any] = {}
    if args.difficulty:
        start_body["difficulty"] = args.difficulty
    start = _request(args.base, "POST", "/api/agents/me/qualify", api_key=key, body=start_body)
    qual_id = start.get("qualMatchId")
    if not qual_id:
        _print({"step": "qualify-start", "register": reg, "response": start})
        return 2

    rounds = []
    qual_status = "IN_PROGRESS"
    for r in range(1, args.max_rounds + 1):
        move = _choose_move(args.strategy, r)
        path = f"/api/agents/me/qualify/{qual_id}/rounds/{r}"
        rr = _request(args.base, "POST", path, api_key=key, body={"move": move})
        rounds.append(rr)
        qual_status = str(rr.get("qualStatus", "IN_PROGRESS")).upper()
        if qual_status in ("PASSED", "FAILED"):
            break

    if qual_status != "PASSED":
        _print({
            "register": reg,
            "qualify": {"qualMatchId": qual_id, "result": qual_status, "rounds": rounds},
            "queue": None,
            "message": "Qualification not passed; not joining queue.",
        })
        return 3

    join = _request(args.base, "POST", "/api/queue", api_key=key)
    _print({
        "register": reg,
        "qualify": {"qualMatchId": qual_id, "result": qual_status, "rounds": rounds},
        "queue": join,
    })
    return 0


def cmd_watch_lobby(args: argparse.Namespace) -> int:
    print(f"Watching {args.base} every {args.interval}s (Ctrl+C to stop)")
    try:
        while True:
            q = _request(args.base, "GET", "/api/queue")
            m = _request(args.base, "GET", "/api/matches")
            snap = {
                "ts": int(time.time()),
                "queue_total": q.get("total", len(q.get("queue", [])) if isinstance(q, dict) else 0),
                "queue": q.get("queue", []),
                "running_matches": [x for x in m.get("matches", []) if x.get("status") == "RUNNING"],
                "recent_matches": [x for x in m.get("matches", []) if x.get("status") == "FINISHED"][:5],
            }
            _print(snap)
            time.sleep(args.interval)
    except KeyboardInterrupt:
        return 0


def build_parser() -> argparse.ArgumentParser:
    p = argparse.ArgumentParser(description="Agent Arena CLI")
    p.add_argument("--base", default=DEFAULT_BASE, help=f"API base URL (default: {DEFAULT_BASE})")

    sub = p.add_subparsers(dest="cmd", required=True)

    s = sub.add_parser("register", help="Register a new bot")
    s.add_argument("--name", required=True)
    s.add_argument("--email")
    s.add_argument("--description")
    s.set_defaults(func=cmd_register)

    s = sub.add_parser("qual-start", help="Start qualification")
    s.add_argument("--key", required=True)
    s.add_argument("--difficulty", choices=["easy", "medium", "hard"], default="easy")
    s.set_defaults(func=cmd_qual_start)

    s = sub.add_parser("qual-round", help="Submit one qualification round")
    s.add_argument("--key", required=True)
    s.add_argument("--qual-id", required=True)
    s.add_argument("--round-no", type=int, required=True)
    s.add_argument("--move", choices=["ROCK", "PAPER", "SCISSORS", "rock", "paper", "scissors"], required=True)
    s.set_defaults(func=cmd_qual_round)

    s = sub.add_parser("qual-auto", help="Auto-play qualification rounds")
    s.add_argument("--key", required=True)
    s.add_argument("--difficulty", choices=["easy", "medium", "hard"], default="easy")
    s.add_argument("--strategy", choices=["paper", "rock", "scissors", "random", "cycle"], default="paper")
    s.add_argument("--max-rounds", type=int, default=5)
    s.set_defaults(func=cmd_qual_auto)

    s = sub.add_parser("join", help="Join queue")
    s.add_argument("--key", required=True)
    s.set_defaults(func=cmd_join)

    s = sub.add_parser("queue", help="Public queue snapshot")
    s.set_defaults(func=cmd_queue)

    s = sub.add_parser("queue-me", help="My queue status")
    s.add_argument("--key", required=True)
    s.set_defaults(func=cmd_queue_me)

    s = sub.add_parser("onboard", help="Register + qualify + join queue")
    s.add_argument("--name", required=True)
    s.add_argument("--email")
    s.add_argument("--description")
    s.add_argument("--difficulty", choices=["easy", "medium", "hard"], default="easy")
    s.add_argument("--strategy", choices=["paper", "rock", "scissors", "random", "cycle"], default="paper")
    s.add_argument("--max-rounds", type=int, default=5)
    s.set_defaults(func=cmd_onboard)

    s = sub.add_parser("watch-lobby", help="Poll queue and matches")
    s.add_argument("--interval", type=int, default=5)
    s.set_defaults(func=cmd_watch_lobby)

    return p


def main() -> int:
    parser = build_parser()
    args = parser.parse_args()
    try:
        return int(args.func(args))
    except ApiException as e:
        _print({"status": e.status, "error": e.payload})
        return 10
    except Exception as e:  # noqa: BLE001
        _print({"error": "CLI_RUNTIME_ERROR", "message": str(e)})
        return 11


if __name__ == "__main__":
    raise SystemExit(main())
